payload/reggae/backend/ninja.d

```diff
@@ -166,179 +166,59 @@ private:
-                              in string last = "", in string after = "") @trusted pure const {
-
-        import std.array: empty;
-        import std.algorithm.searching: canFind;
-
-        auto cmdLine = "command = " ~ targetRawCommand(target);
-        if(!before.empty) cmdLine ~= " $before";
-        cmdLine ~= shellCommand.canFind(" " ~ first) ? " " ~ first : first;
-        if(!between.empty) cmdLine ~= " $between";
-        cmdLine ~= shellCommand.canFind(" " ~ last) ? " " ~ last : last;
-        if(!after.empty) cmdLine ~= " $after";
-
-        return cmdLine;
-    }
-
-    //Ninja operates on rules, not commands. Since this is supposed to work with
-    //generic build systems, the same command can appear with different parameter
-    //ordering. The first time we create a rule with the same name as the command.
-    //The subsequent times, if any, we append a number to the command to create
-    //a new rule
-    string getRuleName(string cmd, in string ruleCmdLine, out bool haveToAdd) @safe nothrow {
-        import std.algorithm.searching: canFind, startsWith;
-        import std.algorithm.iteration: filter;
-        import std.array: array, empty, replace;
-        import std.conv: text;
-
-        immutable ruleMainLine = "rule " ~ cmd;
-        //don't have a rule for this cmd yet, return just the cmd
-        if(!ruleEntries.canFind!(a => a.mainLine == ruleMainLine)) {
-            haveToAdd = true;
-            return cmd;
+        if (i_out < 0) {
+            ruleName = "_custom_in";
+            addParamLine("before", 0, i_in);
+            addParamLine("after", i_in + 3, -1);
+        } else if (i_in < 0) {
+            ruleName = "_custom_out";
+            addParamLine("before", 0, i_out);
+            addParamLine("after", i_out + 4, -1);
+        } else if (i_in < i_out) {
+            ruleName = "_custom_in_out";
+            addParamLine("before", 0, i_in);
+            addParamLine("between", i_in + 3, i_out);
+            addParamLine("after", i_out + 4, -1);
+        } else {
+            ruleName = "_custom_out_in";
+            addParamLine("before", 0, i_out);
+            addParamLine("between", i_out + 4, i_in);
+            addParamLine("after", i_in + 3, -1);
         }
 
-        //so we have a rule for this already. Need to check if the command line
-        //is the same
-
-        //same cmd: either matches exactly or is cmd_{number}
-        auto isSameCmd = (in NinjaEntry entry) {
-            bool sameMainLine = entry.mainLine.startsWith(ruleMainLine) &&
-                (entry.mainLine == ruleMainLine || entry.mainLine[ruleMainLine.length] == '_');
-            bool sameCmdLine = entry.paramLines == [ruleCmdLine];
-
-            return sameMainLine && sameCmdLine;
-        };
+        const includeImplicitInputs = (i_out >= 0);
+        // TODO: weird inputsOverride for implicitInput case?!
+        const buildLine = buildLine(target, ruleName, includeImplicitInputs);
```

Comment:
`implicitInputRule()` used to do some 'weird' stuff wrt. patching the command and preparing some `inputsOverride`, no idea why that was necessary and whether it still is.

I also don't know why the implicit inputs are only included if the custom command has a `$out`.

Reply:
Don't ask me! If the tests pass...